accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -616,6 +609,442 @@ It is a warning to use a null aware operator (`?.`, `?[]`, `?..`, `??`, `??=`, o
+        return;
+      } else {
+        x = null;
+        return;
+      } // x is inferred as int?
+    } else {
+      x = 3;
+    } // x is inferred as int?
+  }
+
+  {
+    var z; // z is an untyped variable
+    if (b) {
+      z = 3; // z is promoted to int
+     } else {
+      z = 3.0; // z is promoted to double
+      return;
+    } // Error, z has no consistent inferred type
+  }
+
+  {
+    var z; // z is an untyped variable
+    if (b) {
+      z = 3; // z is promoted to int
+      return;
+     } else {
+      z = 3.0; // z is promoted to double
+      return;
+    } // Error, z has no consistent inferred type
+  }
+
+  {
+    var u; // u is an untyped variable
+    if (b) {
+      u = 0 as num; // u is promoted to num
+      if (u is! int) u = 1;
+      // u has inferred type num, but is promoted to int
+    } else {
+      u = 3;
+      // u is inferred as int
+    } Error: u has no inferred type
+  }
+}
+```
+
+It is an error to read an **untyped variable** when it has no **initialization
+inferred type**.
+
+It is an error to write to an **untyped variable** when it has no
+**initialization inferred type**, unless the write is an **initializing
+assignment**.
+
+
+```dart
+void test(bool b) {
+  {
+    var x; // x is an untyped variable
+    print(x); // Error, no initialization inferred type
+  }
+
+  {
+    late var x; // x is an untyped variable
+    var f = () => print(x); // Error, x has no initialization inferred type
+  }
+
+  {
+    var x; // x is an untyped variable
+    if (b) {
+      print(x); // Error: x has no initialization inferred type.
+      x = 3;
+    } else {
+      x = 4;
+    } // x is inferred as type int
+    print(x); // No error here, since x has been inferred at this point
+  }
+
+  {
+    var z;
+    do {
+      // z is potentially assigned.
+      if (b) {
+        z = 42;
+      } else {
+        z = 37;
+      }
+    } while (something);
+    print(z); // Error, z has no initialization inferred type
+  }
+```
+
+Notice that the above definitions never discard an **initialization inferred
+type** from any path, and always ensure that the **initialization inferred
+type** is consistent across joins.  Any variable which is given an
+**initialization inferred type** in this way may be treated by IDEs and tools as
+having this type for the purposes of documentation and error messages. An
+untyped cariable which is never read nor written may have no **initialization
+inferred type**.  Such a degenerate variable may be treated by the tools as
+having type `Object?` for the purposes of documentation and error messages.
+
+```dart
+void test(b) {
+  var x; // x is an untyped variable, documented as having type int?
+  var z; // z is an untyped variable, documented as having type Object?
+  if (b) {
+    x = 3; // x is inferred to int
+    return;
+  } else if (b) {
+    return;  // Note, x has no inferred type on this path
+  } else {
+    // x not assigned
+  } // x has type `int?`
+}
+```
+
```

Comment:
Do we need to say that when the specification refers to the *declared type* of a variable, an untyped variable uses its IIT instead, if it has one (and it's probably an error if it doesn't have one).

Reply:
+1.

Side note: I think the spec has similar problems with static and instance variables.  E.g. the "variables" section of the spec says:
> Let v be variable declared in an initializing variable declaration, and let e be the associated initializing expression.  It is a compile-time error if the static type of e is not assignable to the declared type of v.

But if v is an instance variable, and the declaration of v doesn't include a type, and the type is inferred from a base class, then we want this rule to apply to the inferred type of v.

I wonder if it would be better to replace all instances of the term "declared type" in the spec with some other term, and define that term explicitly?